Remove TimePicker state and Moment dependency

pablinos opened the pull request:
<!-- Learn the overall process and best practices for pull requests at https://github.com/WordPress/gutenberg/blob/master/docs/contributors/repository-management.md#pull-requests. -->

## Description
While working on the post schedule feature, @retrofox and I noticed that there are a number of bugs with the implementation of the `DateTimePicker`. The `TimePicker` was holding its own local state, and would only update the month when the select component lost focus. 

This change updates the `TimePicker` component to:

- Not rely on local state
- Use `date-fns` rather than Moment
- Tweak the triggering of the `onChange` callback. 

## How has this been tested?

Locally using Storybook

## Screenshots <!-- if applicable -->

## Types of changes

This could be a breaking change, depending on how others are using the `TimePicker` component.

## Checklist:
- [ ] My code is tested.
- [ ] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/javascript/ -->
- [ ] My code follows the accessibility standards. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/accessibility-coding-standards/ -->
- [ ] My code has proper inline documentation. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/javascript/ -->
- [ ] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR. <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/master/docs/contributors/native-mobile.md -->

retrofox committed:
date-time: highlight site posts

retrofox committed:
post-schedule: pick up events filtering by month

retrofox committed:
post-schedule: replace HOC by hooks

retrofox committed:
post-schedule: refactor getting events by month

retrofox committed:
packakes: add react-datepicker component

retrofox committed:
date-time: replace react-dates by react-datepicker

retrofox committed:
date-time: handling onChange

retrofox committed:
base-styles: define $gray-800

retrofox committed:
date-time: adjust month view styles

retrofox committed:
date-time: restore TimePicker component

retrofox committed:
date-time: add DatePicker to storybook

retrofox committed:
date-time: simplify storybook

retrofox committed:
date-time: improve date picker header

retrofox committed:
date-time: improve date storybook. add set today

retrofox committed:
date-time: handling locale

pablinos committed:
Make the Timepicker a pure component

pablinos committed:
Add date-fns package

retrofox committed:
date-time: start to mark events in calendar

pablinos committed:
Update packages, so DateTimePicker still works

pablinos committed:
Use 'element' variable instead of recalculating the value

pablinos committed:
Adapt the DateTimePicker to use the new TimePicker implementation

pablinos commented on packages/components/src/date-time/index.js:
```diff
@@ -35,7 +35,7 @@ export class DateTimePicker extends Component {
 	render() {
 		const {
-			currentDate,
+			currentDate = new Date(),
```
I'm not sure this default is needed, as it's initialised elsewhere, and should probably be handled by default props instead. This is a reminder to me to remove it!

ZebulanStanphill commented on packages/components/src/date-time/date.js:
```diff
@@ -1,150 +1,53 @@
- * WordPress dependencies
- */
-import { Component, createRef } from '@wordpress/element';
+import set from 'date-fns/set';
+import format from 'date-fns/format';
```
Correct me if I'm wrong, but couldn't this be:

```suggestion
import { format, set } from 'date-fns';
```

pablinos committed:
Use local state for IntegerValidatedField to better handle focus

When using `defaultValue` the change to the DOM node meant that focus
was being lost as the `TimePicker` was rendered.

pablinos commented:
I'm getting a problem when this is used with our version of the `DatePicker` based on React DateTimePicker. It has code to grab focus when the selected date changes. This is happening as I tab out of one of the `TimePicker`'s input fields, but strangely not when I hit enter. This is odd as it causes the same update function to be called.

ZebulanStanphill commented on packages/components/src/date-time/time.js:
```diff
@@ -3,16 +3,18 @@
 import classnames from 'classnames';
 import { isInteger } from 'lodash';
-import moment from 'moment';
+import startOfMinute from 'date-fns/startOfMinute';
+import set from 'date-fns/set';
+import format from 'date-fns/format';
```
```suggestion
import { format, set, startOfMinute } from 'date-fns';
```

ZebulanStanphill commented on packages/components/src/date-time/time.js:
```diff
@@ -26,19 +28,27 @@ import TimeZone from './timezone';
  */
-function UpdateOnBlurAsIntegerField( { as, value, onUpdate, ...props } ) {
-	function handleBlur( event ) {
+function IntegerValidatedField( { as, value, onUpdate, ...props } ) {
+	const element = as ?? 'input';
```
```suggestion
function IntegerValidatedField( { as = 'input', value, onUpdate, ...props } ) {
	const element = as;
```

retrofox committed:
date-time: starting to work on event tooltips

retrofox committed:
date-time: limit events per day

retrofox committed:
packages: add date-fns dependency

retrofox committed:
post-schedule: remove console.log

retrofox committed:
post-schedule: rendering site events

retrofox committed:
date-picker: applying styles to events

retrofox committed:
post=schedule: fix popiver button element

retrofox committed:
post-schedule: limit requestng entities

retrofox committed:
colors: remove $gray-800 color

pablinos committed:
Tidy up imports and fix the linting errors

pablinos committed:
Fix the key warning when rendering the Tooltips

stokesman commented:
I'd wager that all the problems this PR intended to address are since resolved and the last push here is going to be three years old soon. We should close this one out, no?

pablinos commented:
> I'd wager that all the problems this PR intended to address are since resolved and the last push here is going to be three years old soon. We should close this one out, no?

Yes we should!
